src/Fhir.Anonymizer.Core/PartitionedExecution/FhirPartitionedExecutor.cs

```diff
@@ -0,0 +1,113 @@
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException();
+                }
+
+                batchData.Add(content);
+                if (batchData.Count < BatchSize)
+                {
+                    continue;
+                }
+
+                executionTasks.Enqueue(AnonymizeAsync(batchData, breakOnAnonymizationException, progress, cancellationToken));
+                batchData = new List<string>();
+                if (executionTasks.Count < PartitionCount)
+                {
+                    continue;
+                }
+
+                await ConsumeExecutionResultTask(executionTasks).ConfigureAwait(false);
+            }
+
+            if (batchData.Count > 0)
+            {
+                executionTasks.Enqueue(AnonymizeAsync(batchData, breakOnAnonymizationException, progress, cancellationToken));
+            }
+
+            while (executionTasks.Count > 0)
+            {
+                await ConsumeExecutionResultTask(executionTasks).ConfigureAwait(false);
+            }
+
+            await AnonymizedDataConsumer.CompleteAsync().ConfigureAwait(false);
+        }
+
+        private async Task<IEnumerable<string>> AnonymizeAsync(List<string> batchData, bool breakOnAnonymizationException, IProgress<BatchAnonymizeProgressDetail> progress, CancellationToken cancellationToken)
+        {
+            return await Task.Run(() =>
+            {
+                List<string> result = new List<string>();
+
+                BatchAnonymizeProgressDetail batchAnonymizeProgressDetail = new BatchAnonymizeProgressDetail();
+                foreach (string content in batchData)
+                {
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw new OperationCanceledException();
+                    }
+
+                    try
+                    {
+                        string anonymizedResult = AnonymizerFunction(content);
+                        result.Add(anonymizedResult);
+                        batchAnonymizeProgressDetail.Completed++;
+                    }
+                    catch (Exception)
+                    {
+                        if (breakOnAnonymizationException)
+                        {
+                            throw;
+                        }
+
+                        batchAnonymizeProgressDetail.Failed++;
+                    }                    
+                }
+
+                progress?.Report(batchAnonymizeProgressDetail);
+                return result;
+            }).ConfigureAwait(false);
+        }
+
+        private async Task ConsumeExecutionResultTask(Queue<Task<IEnumerable<string>>> executionTasks)
+        {
+            IEnumerable<string> resultContents = await executionTasks.Dequeue().ConfigureAwait(false);
+
+            await AnonymizedDataConsumer.ConsumeAsync(resultContents).ConfigureAwait(false);
```

Comment:
what is the time taken by anonymize vs disk writing? I am wondering if its worthwhile to avoid waiting on IO here.